full-service/src/service/memo.rs

```diff
@@ -0,0 +1,123 @@
+use crate::{
+    db::{
+        account::{AccountID, AccountModel},
+        models::Account,
+        WalletDbError,
+    },
+    service::ledger::{LedgerService, LedgerServiceError},
+    util::b58::{b58_decode_public_address, B58Error},
+    WalletService,
+};
+use mc_account_keys::AccountKey;
+use mc_connection::{BlockchainConnection, UserTxConnection};
+use mc_crypto_keys::{KeyError, RistrettoPublic};
+use mc_fog_report_validation::FogPubkeyResolver;
+use mc_transaction_extra::{MemoDecodingError, MemoType, UnusedMemo};
+use mc_util_serial::DecodeError;
+use std::{convert::TryFrom, ops::DerefMut};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+#[allow(clippy::large_enum_variant, clippy::result_large_err)]
+pub enum MemoServiceError {
+    #[error("WalletDb: {0}")]
+    WalletDb(WalletDbError),
+
+    #[error("B58: {0}")]
+    B58(#[from] B58Error),
+
+    #[error("Decode: {0}")]
+    Decode(DecodeError),
+
+    #[error("LedgerService: {0}")]
+    LedgerService(LedgerServiceError),
+
+    #[error("Key: {0}")]
+    Key(KeyError),
+
+    #[error("MemoDecoding: {0}")]
+    MemoDecoding(MemoDecodingError),
+
+    #[error("Invalid memo type for validation, expecting AuthenticatedSenderMemo.")]
+    InvalidMemoTypeForValidation,
+}
+
+impl From<WalletDbError> for MemoServiceError {
+    fn from(e: WalletDbError) -> Self {
+        Self::WalletDb(e)
+    }
+}
+
+impl From<DecodeError> for MemoServiceError {
+    fn from(e: DecodeError) -> Self {
+        Self::Decode(e)
+    }
+}
+
+impl From<LedgerServiceError> for MemoServiceError {
+    fn from(e: LedgerServiceError) -> Self {
+        Self::LedgerService(e)
+    }
+}
+
+impl From<KeyError> for MemoServiceError {
+    fn from(e: KeyError) -> Self {
+        Self::Key(e)
+    }
+}
+
+impl From<MemoDecodingError> for MemoServiceError {
+    fn from(e: MemoDecodingError) -> Self {
+        Self::MemoDecoding(e)
+    }
+}
```

Comment:
IIUC we can replace these `impl`s with `#[from]` in the `enum` declaration

Reply:
So, it requires that the error we are going from conforms to std::Error, otherwise this is still required. The goal is to make as many things conform to that as possible so we can use the #[from] macro, but that will be a later PR that is focused on that.

Reply:
no longer relevant, will be address thiserror in a different PR potentially